src/openqaoa-core/tests/test_workflows.py

```diff
@@ -1726,6 +1731,640 @@ def test_change_properties_after_compilation(self):
+            fqaoa.set_device(create_device(location="local", name=d))
+            assert type(fqaoa.device) == DeviceLocal
+            assert fqaoa.device.device_name == d
+            assert fqaoa.device.device_location == "local"
+
+    def test_compile_before_optimise(self):
+        """
+        Assert that compilation has to be called before optimisation
+        """
+
+        fqaoa = FQAOA()
+        fqaoa.set_classical_optimizer(optimization_progress=True)
+
+        self.assertRaises(ValueError, lambda: fqaoa.optimize())
+
+    def test_cost_hamil(self):
+        num_assets, budget = 5, 3
+        mu, sigma, _ = generate_random_portfolio_data(num_assets = num_assets, num_days = 15, seed = 1)
+        qubo_problem = PortfolioOptimization(mu, sigma, risk_factor = None, budget = budget, penalty = None).qubo
+
+        test_hamil = Hamiltonian.classical_hamiltonian(
+            terms=qubo_problem.terms,
+            coeffs=qubo_problem.weights,
+            constant=qubo_problem.constant,
+        )
+
+        fqaoa = FQAOA()
+
+        fqaoa.fermi_compile(problem=qubo_problem, n_fermions=budget)
+
+        self.assertEqual(fqaoa.cost_hamil.expression, test_hamil.expression)
+        self.assertEqual(
+            fqaoa.qaoa_descriptor.cost_hamiltonian.expression, test_hamil.expression
+        )
+
+    def test_set_circuit_properties_fourier_q(self):
+        """
+        The value of q should be None if the param_type used is not fourier.
+        Else if param_type is fourier, fourier_extended or fourier_w_bias, it
+        should be the value of q, if it is provided.
+        """
+
+        fourier_param_types = ["fourier", "fourier_extended", "fourier_w_bias"]
+
+        fqaoa = FQAOA()
+
+        for each_param_type in fourier_param_types:
+            fqaoa.set_circuit_properties(param_type=each_param_type, q=1)
+            self.assertEqual(fqaoa.circuit_properties.q, 1)
+
+        fqaoa.set_circuit_properties(param_type="standard", q=1)
+
+        self.assertEqual(fqaoa.circuit_properties.q, None)
+
+    def test_set_circuit_properties_annealing_time_linear_ramp_time(self):
+        """
+        Check that linear_ramp_time and annealing_time are updated appropriately
+        as the value of p is changed.
+        """
+
+        fqaoa = FQAOA()
+
+        fqaoa.set_circuit_properties(p=3)
+
+        self.assertEqual(fqaoa.circuit_properties.annealing_time, 0.7 * 3)
+        self.assertEqual(fqaoa.circuit_properties.linear_ramp_time, 0.7 * 3)
+
+        fqaoa.set_circuit_properties(p=2)
+
+        self.assertEqual(fqaoa.circuit_properties.annealing_time, 0.7 * 2)
+        self.assertEqual(fqaoa.circuit_properties.linear_ramp_time, 0.7 * 2)
+
+    def test_set_circuit_properties_qaoa_descriptor_mixer_x(self):
+        """
+        Checks if the X mixer created by the X_mixer_hamiltonian method
+        and the automated methods in workflows do the same thing.
+
+        For each qubit, there should be 1 RXGateMap per layer of p.
```

Comment:
Also X mixer shouldn't be allowed in combination with `FQAOA`, right?

Reply:
Yes, you are right.

Fix docstring and code.